docs/architecture/engine.md

```diff
@@ -0,0 +1,197 @@
+
+If the database is stopped for any reason in-between "flushes" (writing fileset files out to disk), then when the node is started back up those writes will need to be recovered by reading the commitlog or streaming in the data from a peer responsible for the same shard (if the replication factor is larger than 1.)
+
+The blocksize parameter is the most important variable that needs to be tuned for your workload. A small blocksize will mean more frequent flushing and a smaller memory footprint for the data that is being actively compressed, but it will also reduce the compression ratio and your data will take up more space on disk.
+
+While the [fileset files](storage.md) are designed to support efficient data retrieval via the series primary key (the ID), there is still a heavy cost associated with any query that has to retrieve data from disk. To compensate for that, M3DB support various [caching policies](caching.md) which can significantly improve the performance of reads.
+
+## Lifecycle of a write
+
+We now have enough context of M3DB's architecture to discuss the lifecycle of a write. A write begins when an M3DB client calls the `writeBatchRaw` endpoint on M3DB's thrift server. The write itself will contain the following information:
+
+1. The namespace
+2. The series ID (byte blob)
+3. The timestamp
+4. The value itself
+
+M3DB will consult the database object to check if the namespace exists, and if it does,then it will hash the series ID to determine which shard it belongs to. If the node receiving the write owns that shard, then it will lookup the series in the shard object. If the series exists, then it will lookup the series corresponding encoder and encode the datapoint into the compressed stream. If the encoder doesn't exist (no writes for this series have occurred yet as part of this block) then a new encoder will be allocated and it will begin a compressed M3TSZ stream with that datapoint. There is also some special logic for handling out-of-order writes which is discussed in the [merging duplicate encoders section](engine.md#merging-duplicate-enoders).
+
+At the same time, the write will be appended to the commitlog queue (and depending on the commitlog configuration immediately fsync'd to disk or batched together with other writes and flushed out all at once).
+
+The write will exist only in this "active buffer" and the commitlog until the block ends and is flushed to disk, at which point the write will exist in a fileset file for efficient storage and retrieval later and the commitlog entry can be garbage collected.
+
```

Comment:
Need to further describe mulit-node fanout/consistency guarantees the client will ensure before returning success/failure.